test-support/helpers/upload.js

```diff
@@ -1,19 +1,36 @@
-/*global triggerEvent, find */
+/* global global, upload, window */
+import Ember from 'ember';
 
-export default function (selector, file, filename) {
-  let input = findWithAssert(selector)[0];
+Ember.Test.registerAsyncHelper('upload', (app, selector, file, filename) => {
+  const { findWithAssert, triggerEvent } = app.testHelpers;
+
+  const input = findWithAssert(selector)[0];
 
   file.name = filename;
 
   // This hack is here since we can't mock out the
   // FileList API easily; we're taking advantage
   // that we can mutate the FileList DOM API at
   // runtime to allow us to push files into the <input>
-  let files = [file];
-  input.files.item = function (idx) {
+  const files = [file];
+
+  input.files.item = function(idx) {
     return files[idx];
   };
+
   input.files.size = files.length;
 
   return triggerEvent(selector, 'change');
+});
+
+export default function(selector, file, filename) {
+  const scope = window || global;
+  const fakeApplicationFromScope = {
+    testHelpers: {
+      findWithAssert: scope.findWithAssert,
+      triggerEvent: scope.triggerEvent,
+    }
+  };
+
+  return upload(fakeApplicationFromScope, selector, file, filename);
```

Comment:
This exported function is an _attempt_ to maintain backward compatibility with client applications that are `import`-ing the `upload` helper.

I'm not sure if this is a good pattern to adopt, or even if it'll work, but the idea is that since `triggerEvent` and `findWithAssert` were previously invoked from the global scope, we should pass those in so that they'll continue to be called that way.

What do you think?

Reply:
I think that's a good idea here- I would prefer to have something else cause a major version release, but if it's necessary, it's necessary.